utils/build-script

```diff
@@ -774,6 +778,115 @@ class BuildScriptInvocation(object):
+
+        product_classes = []
+        product_classes.append(products.CMark)
+        product_classes.append(products.LLVM)
+        product_classes.append(products.Clang)
+        if self.args.build_lldb:
+            product_classes.append(products.LLDB)
+        if self.args.build_llbuild:
+            product_classes.append(products.LLBuild)
+        if self.args.build_libdispatch:
+            product_classes.append(products.LibDispatch)
+        if self.args.build_foundation:
+            product_classes.append(products.Foundation)
+        if self.args.build_xctest:
+            product_classes.append(products.XCTest)
+        if self.args.build_swiftpm:
+            product_classes.append(products.SwiftPM)
+        return product_classes
+
+    def execute(self):
+        """Execute the invocation with the configured arguments."""
+
+        # Convert to a build-script-impl invocation.
+        (impl_env, impl_args) = self.convert_to_impl_arguments()
+
+        # If using the legacy implementation, delegate all behavior to
+        # `build-script-impl`.
+        if self.args.legacy_impl:
+            # Execute the underlying build script implementation.
+            call_without_sleeping([build_script_impl] + impl_args,
+                                  env=impl_env, echo=True)
+            return
+
+        # Otherwise, we compute and execute the individual actions ourselves.
+
+        def execute_one_impl_action(host=None, product_class=None, name=None):
+            if host is None:
+                assert (product_class is None and
+                        name == "merged-hosts-lipo"), "invalid action"
+                action_name = name
+            elif product_class is None:
+                assert name == "package", "invalid action"
+                action_name = "{}-{}".format(host.name, name)
+            else:
+                assert name is not None, "invalid action"
+                action_name = "{}-{}-{}".format(
+                    host.name, product_class.product_name(), name)
+            call_without_sleeping(
+                [build_script_impl] + impl_args + [
+                    "--only-execute", action_name],
+                env=impl_env, echo=self.args.verbose_build)
+
+        # Compute the list of hosts to operate on.
+        all_host_names = [
+            self.args.host_target] + self.args.cross_compile_hosts
+        all_hosts = [StdlibDeploymentTarget.get_target_for_name(name)
+                     for name in all_host_names]
+
+        # Compute the list of product classes to operate on.
+        #
+        # FIXME: This should really be per-host, but the current structure
+        # matches that of `build-script-impl`.
+        product_classes = self.compute_product_classes()
+
+        # Execute each "pass".
+
+        # Build...
+        for host_target in all_hosts:
+            # FIXME: We should only compute these once.
+            config = HostSpecificConfiguration(host_target.name, self)
```

Comment:
I basically agree with all your points, but the problem is that it is very hard to enact cleanups when the implementation is split between the Python code and the bash code. I really think we are better off getting the migration done and then continuing to clean up the code.

More specific comments:

1. The reason I didn't put HostSpecificConfiguration into its own file was because its logic was so intimately tied to the current `args` set (which we agree is insufficient). I didn't want to have a constructor in `swift_build_support` that just used a grab bag of `args` items, and I also didn't want to try and explode all of the things it currently uses into separate arguments (or its own object). I'm not sure how to resolve this difficulty currently, but agree it would be good to have something better and have tests. Do you have ideas?

2. I agree about making everything store more structured objects instead of names, and will get to that (eventually). I'm happy to have the #2835 version of this.

3. I agree we should have an object that represents the parsed arguments better than the existing invocation, and that object should live inside `swift_build_support`. Having `Invocation` be the sole object which understand the "command line arguments" and turns that into structured information makes sense to me though, and matches a convention we use inside the Clang and Swift compilers.